High amount of new features and fixes

Givikap120 opened the pull request:
1) Draw frames on a timeline option
2) Draw hitwindows on a timeline option
3) Draw sliderend range option
4) Sliderends is now drawn
5) Moving frames in time
6) Adding new frames
7) Centering frames
8) Scaling Frames
9) Deleting Frames
10) Changing all inputs to keyboard button
11) (not accurate) missed sliderend detection
12) Changing map file of the replay while keeping inputs
13) Deselect active button by LMB feature
14) Deleted 2.00x playback speed (pretty useless) and added 0.05x playback speed instead
15) Changed time variables that are not bound to frames and objects to double (fix some problems)
16) Added doubletap detection in accuracy analyser

Givikap120 committed:
Gigantic numbers of feautres added

Givikap120 committed:
Merge branch 'master' of https://github.com/Givikap120/osuReplayEditorV3_SRC

Givikap120 commented:
There is some bug where scrubing right is much faster than scrubing left. Especially noticable on 0.05x speed but happens an all speeds.
I have no idea what is the reason of this.

Givikap120 committed:
Fixed different scrub speed problem

Givikap120 commented:
Also, many very hard to fix things like correct accuracy calculation (what is wrong even in rewind, only danser get it right) is easily fixable with using lazer framework. Idk why you didn't used it in first place.

thebetioplane commented:
I haven't looked at all the code yet. You should probably split up your changes into smaller PRs in the future which would make it easier to review changes.

> Also, many very hard to fix things like correct accuracy calculation (what is wrong even in rewind, only danser get it right) is easily fixable with using lazer framework. Idk why you didn't used it in first place.

The reason why I am not using lazer framework is the same reason I have the UI in C# and the core logic in C++, the same reason I draw using OpenGL 1.2,  the same reason why multiple coding styles are mixed, and the same reason why the replay editor does not have proper error handling: Because this code is old and not really well maintained by me. The core code is a mixture of stuff I did a long time ago when I was a foolish programmer combined with stuff I did more recently. I don't spend that much time programming this at all really anymore though, as is evident by the commit history. You have to understand that I write software as a full time job, so when in my free time I have little interest to do more of it.

Creating a custom ruleset (? Don't know if that is the correct name ?) in lazer is the correct way to make a replay editor if I was to do it again. It is something that I thought about a couple years ago. If there was to be an osuReplayEditorV4, or if I were to recommend you make a replay editor, I would choose that option. There is no reason not to. It will allow you to draw all the hit objects without any trouble and support both legacy and lazer beatmap and replay formats. I don't actually think it would be all that complicated really.

Givikap120 commented:
Ahh, separating it into different PRs would be tough - because i've added all features that i've need one by one in one project

thebetioplane commented:
I was briefly glancing at this and overall it looks good. I haven't manually tested everything though (and there are no auto tests) so I still need to do that. I did have a question about this:

> Changed time variables that are not bound to frames and objects to double (fix some problems)

Which bugs did this fix?

thebetioplane commented on ReplayEditor/API.cpp:
```diff
@@ -149,6 +149,25 @@ DLLFUN(BOOL) Cleanup()
+    return ptr;
+}
+
+DLLFUN(void) FreeBuffer(void* buffer)
```
I don't like the idea of passing unmanaged memory to C#. Check MetadataEditor/MetadataEditor.cs line 208 for an example. That method makes the C# manage the memory and reduces risk of leak error.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/API.cpp:
```diff
@@ -180,6 +200,19 @@ DLLFUN(BOOL) LoadReplay(const wchar_t *fname)
+{
+    using ::replayengine::Replay;
+    std::wstring filename(fname);
+    filename.push_back('\0');
```
What is the purpose of appending a `\0`?

Givikap120 replied:
it says that filenames are different in hashmap otherwise

thebetioplane commented on ReplayEditor/API.cpp:
```diff
@@ -278,7 +328,12 @@ DLLFUN(void) JumpTo(SongTime_t ms)
 }
 
-DLLFUN(void) RelJump(SongTime_t ms)
+//DLLFUN(void) RelJump(SongTime_t ms)
```
Remove commented out code please.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/API.cpp:
```diff
@@ -303,7 +358,12 @@ DLLFUN(float) GetPlaybackSpeed()
 }
 
-DLLFUN(SongTime_t) GetTime()
+//DLLFUN(SongTime_t) GetTime()
```
Remove commented out code please.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/accuracy_analyzer.cpp:
```diff
@@ -58,25 +58,64 @@ void accuracy_analyzer::analyze(Stats* stats, bool do_trace)
-            const bool curr_frame_pressed = curr_frame.pressed_mouse1() || curr_frame.pressed_mouse2();
-            const float ball_radius = beatmapengine::circleradius() * 2.4f;
+
+            double interpolation = double(curr_obj.start - prev_frame.ms) / (curr_frame.ms - prev_frame.ms);
```
The frame points are actually floats (glm::vec2 default), so your interpolation should also be a float not a double.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/accuracy_analyzer.cpp:
```diff
@@ -130,6 +173,41 @@ void accuracy_analyzer::analyze(Stats* stats, bool do_trace)
             }
             ++hitobject_index;
+
+            //if doubletapped
```
```suggestion
            // if doubletapped
```

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/hitobject.cpp:
```diff
@@ -8,13 +8,16 @@
 #define KIND_IS_SLIDER (kind & 2)
 #define KIND_IS_SPINNER (kind & 8)
 
+#define RATIO(a, b) (static_cast<float>(a) / static_cast<float>(b))
```
Isn't `RATIO` already defined in ReplayEditor/stdafx.h?

Givikap120 replied:
fixed

thebetioplane commented on osuReplayEditor/MainForm.cs:
```diff
@@ -880,5 +907,87 @@ private void analyzeAccTraceToolStripMenuItem_Click(object sender, EventArgs e)
+            bool result = API.ChangeHashOfBeatmap(relativePath);
+            if(result)
+            {
+                System.Threading.Thread.Sleep(500);
```
Why is there a need for a sleep here?

Givikap120 replied:
fixed (used to test smth, forgor to remove)

thebetioplane commented on osuReplayEditor/MainForm.cs:
```diff
@@ -880,5 +907,87 @@ private void analyzeAccTraceToolStripMenuItem_Click(object sender, EventArgs e)
+
+        private void MainForm_MouseDown(object sender, MouseEventArgs e)
+        {
+            this.ActiveControl = null;
```
Which problem does this solve?

Givikap120 replied:
it's just deselects the active button when you're pressing on random place
i've added it cuz I constantly pressing space when trying to play the song (habit) and the last action getting repeated

thebetioplane commented on osuReplayEditor/MainForm.cs:
```diff
@@ -590,6 +595,16 @@ private void openToolStripMenuItem_Click(object sender, EventArgs e)
+            if (ptr == null) return;
+
+            string path = Marshal.PtrToStringUni(ptr);
+            API.FreeBuffer(ptr);
```
See other comment I left in API.cpp, I think we should manage the memory in C#.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/slider.hpp:
```diff
@@ -10,6 +10,10 @@
 struct pos_time_t {
     glm::vec2 pos;
     SongTime_t time;
+
```
Is this added constructor used anywhere? Generally with structs it should be unnnecessary to have a constructor because you can always default initalize them with `{}` or value initialize them like `{ .pos = {1.f, 2.f }, .time  = 55 }`

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/osudb.cpp:
```diff
@@ -134,4 +136,12 @@ bool get_entry(const std::string &hash, std::wstring &osu_path, std::wstring &so
     return true;
 }
 
+bool get_hash(const std::wstring &osu_path, std::string &hash)
```
Could you explain more what finding the beatmap hash by file name was meant to solve?

Givikap120 replied:
this is part of changing diff hash of the replay
user selects diff file (.osu) and this searches for it's hash in osu.db

thebetioplane commented on ReplayEditor/audioengine.cpp:
```diff
@@ -30,13 +30,16 @@ class BassAudioEngine : public AudioEngine
     bool is_playing() override;
```
Remove all the commented out code in this file.

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/beatmapengine.cpp:
```diff
@@ -446,15 +466,50 @@ bool beatmapengine::init(const std::wstring &fname)
+    if (start == hitobjects.end()) return;
+    if (end != hitobjects.end()) ++end;
+
+    // glMatrixMode(GL_PROJECTION);
```
Remove this commented out code.

Givikap120 replied:
fixed

thebetioplane commented on osuReplayEditor/MainForm.resx:
```diff
@@ -126,8 +126,11 @@
+  </metadata>
   <metadata name="$this.TrayHeight" type="System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089">
-    <value>25</value>
+    <value>57</value>
```
Do you know what these changes are? I think they come from the designer right, but I couldn't see what they visually changed.

Givikap120 replied:
this is automatic changes by VS, reason is unknown
I didn't turned them off in the PR because it's most likely will break newly added UI elements

thebetioplane commented on ReplayEditor/replayengine.cpp:
```diff
@@ -406,6 +406,248 @@ void Replay::place_marks_at(I64 mark_in, I64 mark_out)
+    if (m_mark_out != m_frames.size() - 1) {
+        SongTime_t initial_delta = m_frames[m_mark_out].ms - m_frames[m_mark_in].ms;
+        SongTime_t max_delta = m_frames[m_mark_out + 1].ms - m_frames[m_mark_in].ms - 1;
+        float maximum_scale = double(max_delta) / initial_delta;
```
```suggestion
        float maximum_scale = static_cast<double>(max_delta) / initial_delta;
```

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/slider.cpp:
```diff
@@ -347,13 +349,22 @@ glm::vec2 slider_t::end_pos() const
     glm::vec2 slider_ball_pos{0.f, 0.f};
     draw_slider_ball = ball_position_at_time(ms, start, end, slider_ball_pos);
+
+    SongTime_t true_slider_end_time = std::max(float(end - last_tick_offset), end - duration() / 2);
```
```suggestion
    SongTime_t true_slider_end_time = std::max(static_cast<float>(end - last_tick_offset), end - duration() / 2.0f);
```

Givikap120 replied:
fixed

thebetioplane commented on ReplayEditor/replayengine.cpp:
```diff
@@ -406,6 +406,248 @@ void Replay::place_marks_at(I64 mark_in, I64 mark_out)
+        if (difference < size) return 0;
+
+        int i = 1;
+        double base = (double)difference / (size + 1);
```
```suggestion
        double base = static_cast<double>(difference) / (size + 1);
```

Givikap120 replied:
honestly I don't understand the meaning of changing cast to c++ style
it looks ugly IMO
i'm using c++ style casting only in dynamic or interpret casts

Givikap120 replied:
fixed

thebetioplane reviewed (commented):
I will publish some initial comments, I am a bit overwhelmed by the amount of changes in this. I apologize I wasn't able to leave comments on all of it yet, but some things I wanted clarification on, and other comments require just small things I noticed while reading your code. I do appreciate your contribution and I will do a more detailed review later.

Also there is a .clang-format file attached to this project. I noticed you left some trailing whitespace and strange line breaks on some of your lines. Just something you can try is formatting it with that. If you don't have clang-format installed or configured (it is bundled with VisualStudio), that's also fine I can format for you later.

Givikap120 commented:
> I was briefly glancing at this and overall it looks good. I haven't manually tested everything though (and there are no auto tests) so I still need to do that. I did have a question about this:
> 
> > Changed time variables that are not bound to frames and objects to double (fix some problems)
> 
> Which bugs did this fix?

10% scrolling is actually not 10% but around 6% due to rounding
and this 6% is the limit of scrolling cuz it's 1ms and can't be any less

Givikap120 committed:
bug fixes and comments removements

now reloads db when it can't find the hash
slider ends misses now not ruin the notelock mechanism

Givikap120 committed:
fix memory managment

Givikap120 committed:
cast fixes

Givikap120 committed:
Update MainForm.cs
